libraries/AC_PID/AC_PID_2D.cpp

```diff
@@ -133,16 +133,14 @@ void AC_PID_2D::update_i(const Vector2f &limit)
     _pid_info_y.limit = false;
 
     Vector2f delta_integrator = (_error * _ki) * _dt;
-    if (!is_zero(limit.length_squared())) {
-        // zero delta_vel if it will increase the velocity error
-        if (is_positive(delta_integrator * limit)) {
-            delta_integrator.zero();
-            _pid_info_x.limit = true;
-            _pid_info_y.limit = true;
-        }
+    float integrator_length = _integrator.length();
+    _integrator += delta_integrator;
+    // do not let integrator increase in length if delta_integrator is in the direction of limit
+    if (is_positive(delta_integrator * limit) && _integrator.limit_length(integrator_length)) {
```

Comment:
You can remove the float by moving it to a direct call in the if. This means that it will only be evaluated if `is_positive(delta_integrator * limit)` is true.  
```suggestion
    if (is_positive(delta_integrator * limit) && _integrator.limit_length(_integrator.length())) {
```

Reply:
No, we need to calculate the length before we add the delta_integrator so we can't do this.

Reply:
yep, your right. I missed that.